File: mlx_sparse/_coo.py
# Copyright (c) 2026 The mlx-sparse contributors - All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from dataclasses import dataclass
from numbers import Number

import mlx.core as mx

import mlx_sparse._native as _native
from mlx_sparse._csr import CSRArray
from mlx_sparse._typing import Shape2D, ValidationMode
from mlx_sparse._validation import (
    ensure_mx_array,
    normalize_shape,
    normalize_validation_mode,
    validate_coo_metadata,
    validate_coo_values,
)


@dataclass(frozen=True, slots=True)
class COOArray:
    """A 2D sparse matrix in Coordinate (COO) format.

    COOArray stores a sparse matrix as three parallel arrays:

    - **data**: non-zero values, shape ``(nnz,)``.
    - **row**: row coordinate for each value, shape ``(nnz,)``.
    - **col**: column coordinate for each value, shape ``(nnz,)``.

    COO is the primary *construction* format. It allows duplicate
    ``(row, col)`` entries and does not require sorted coordinates, making it
    straightforward to assemble matrices from element lists, graph adjacency
    lists, finite-element stencils, or Hamiltonians.

    COO supports native sparse-dense products directly. For heavily repeated
    row-oriented workloads, CSR may still be preferable after construction
    because its compressed row layout avoids coordinate scatter.

    **Format invariants** (checked by ``validate="metadata"`` by default):

    - All three arrays must be rank-1 with the same length.
    - ``row`` and ``col`` share the same integer dtype (``int32`` or
      ``int64``).
    - ``data`` dtype is one of ``float32``, ``float16``, ``bfloat16``, or
      ``complex64``.

    **Additional value-level invariants** (``validate="full"`` only):

    - ``0 <= row[i] < n_rows`` for all entries.
    - ``0 <= col[i] < n_cols`` for all entries.

    Args:
        data: Non-zero values, shape ``(nnz,)``.
        row: Row coordinates, shape ``(nnz,)``.
        col: Column coordinates, shape ``(nnz,)``.
        shape: Matrix dimensions as ``(n_rows, n_cols)``.
        has_canonical_format: Hint that coordinates are sorted and duplicate-
            free. Defaults to ``False``.

    Example::

        import mlx.core as mx
        import mlx_sparse as ms

        data = mx.array([2.0, -1.0, 4.0], dtype=mx.float32)
        row = mx.array([0, 0, 1], dtype=mx.int32)
        col = mx.array([0, 2, 1], dtype=mx.int32)

        # 2×3 matrix:  [[2, 0, -1],
        #               [0, 4,  0]]
        coo = ms.coo_array((data, (row, col)), shape=(2, 3))
        csr = coo.tocsr(canonical=True)
    """

    data: mx.array
    row: mx.array
    col: mx.array
    shape: Shape2D
    has_canonical_format: bool = False

    def __post_init__(self) -> None:
        object.__setattr__(self, "shape", normalize_shape(self.shape))

    @property
    def nnz(self) -> int:
        """Number of stored values (including any duplicates)."""
        return int(self.data.shape[0])

    @property
    def dtype(self):
        """Value dtype of the stored non-zeros (e.g. ``mlx.core.float32``)."""
        return self.data.dtype

    @property
    def index_dtype(self):
        """Integer dtype used for ``row`` and ``col``."""
        return self.row.dtype

    @property
    def ndim(self) -> int:
        """Always 2. Sparse arrays in this package are rank-2."""
        return 2

    def __repr__(self) -> str:
        return (
            "COOArray("
            f"shape={self.shape}, nnz={self.nnz}, dtype={self.dtype}, "
            f"index_dtype={self.index_dtype}, "
            f"has_canonical_format={self.has_canonical_format})"
        )

    def tocsr(self, *, canonical: bool = False) -> CSRArray:
        """Convert to :class:`CSRArray`.

        Sorts entries by row then column and builds a ``(n_rows + 1,)`` row
        pointer array. Duplicate ``(row, col)`` entries are preserved in the
        raw output. Pass ``canonical=True`` to sum them.

        Args:
            canonical: If ``True``, call :meth:`~CSRArray.canonicalize` on the
                result to sort indices and sum duplicates. Default ``False``.

        Returns:
            A :class:`CSRArray` with ``sorted_indices=True``. If
            ``canonical=True``, also ``has_canonical_format=True``.
        """
        data, indices, indptr = _native.coo_tocsr(
            self.data, self.row, self.col, self.shape
        )
        csr = CSRArray(
            data=data,
            indices=indices,
            indptr=indptr,
            shape=self.shape,
            sorted_indices=True,
            has_canonical_format=False,
        )
        if canonical:
            return csr.canonicalize()
        return csr

    def tocsc(self, *, canonical: bool = False):
        """Convert to :class:`~mlx_sparse.CSCArray`."""
        from mlx_sparse._csc import CSCArray

        data, indices, indptr = _native.coo_tocsc(
            self.data, self.row, self.col, self.shape
        )
        csc = CSCArray(
            data=data,
            indices=indices,
            indptr=indptr,
            shape=self.shape,
            sorted_indices=True,
            has_canonical_format=False,
        )
        if canonical:
            return csc.canonicalize()
        return csc

    def todense(self) -> mx.array:
        """Materialize as a dense MLX array.

        Internally converts to CSR and then calls
        :meth:`~CSRArray.todense`. Duplicate entries are summed.

        Returns:
            Dense array of shape ``(n_rows, n_cols)`` with the same dtype as
            ``self.data``.
        """
        return self.tocsr(canonical=False).todense()

    def row_sums(self) -> mx.array:
        """Return the sum of stored values in each COO row."""
        from mlx_sparse._ops import coo_row_sums

        return coo_row_sums(self)

    def col_sums(self) -> mx.array:
        """Return the sum of stored values in each COO column."""
        from mlx_sparse._ops import coo_col_sums

        return coo_col_sums(self)

    def column_sums(self) -> mx.array:
        """Alias for :meth:`col_sums`."""
        return self.col_sums()

    def row_norms(self) -> mx.array:
        """Return the dense-semantics L2 norm of each COO row as ``float32``."""
        from mlx_sparse._ops import coo_row_norms

        return coo_row_norms(self)

    def col_norms(self) -> mx.array:
        """Return the dense-semantics L2 norm of each COO column as ``float32``."""
        from mlx_sparse._ops import coo_col_norms

        return coo_col_norms(self)

    def column_norms(self) -> mx.array:
        """Alias for :meth:`col_norms`."""
        return self.col_norms()

    def diagonal(self) -> mx.array:
        """Return the summed main diagonal."""
        from mlx_sparse._ops import coo_diagonal

        return coo_diagonal(self)

    def trace(self) -> mx.array:
        """Return the summed main diagonal as a scalar."""
        from mlx_sparse._ops import coo_trace

        return coo_trace(self)

    def sum(self, axis=None) -> mx.array:
        """Sum sparse values over all entries, rows, or columns.

        ``axis=None`` returns a scalar, ``axis=1`` returns row sums, and
        ``axis=0`` returns column sums.
        """
        if axis is None:
            return mx.sum(self.row_sums())
        if axis in (1, -1):
            return self.row_sums()
        if axis in (0, -2):
            return self.col_sums()
        raise ValueError(f"COOArray.sum axis must be None, 0, or 1; got {axis!r}.")

    def __matmul__(self, rhs):
        """Matrix multiplication via the ``@`` operator."""
        from mlx_sparse._csc import CSCArray
        from mlx_sparse._ops import coo_matmat, coo_matmul, coo_matvec

        if isinstance(rhs, COOArray):
            return coo_matmat(self, rhs)
        if isinstance(rhs, (CSRArray, CSCArray)):
            raise NotImplementedError(
                "Mixed-format COO sparse-sparse matmul is not implemented. "
                "Convert explicitly if another format is acceptable for your workload."
            )

        rhs = ensure_mx_array(rhs)
        if rhs.ndim == 1:
            return coo_matvec(self, rhs)
        if rhs.ndim >= 2:
            return coo_matmul(self, rhs)
        raise ValueError(f"COO matmul expects rank-1 or higher RHS, got {rhs.shape}.")

    def __rmul__(self, other):
        """Multiply the current CSCArray by a number using the ``*`` operator.

        This returns a new CSCArray with the data multiplied by the number, and
        therefore does not in-place mutate the current CSCArray.

        Args:
            other: A valid number (complex or not).

        Returns:
            A new CSCArray with the data multiplied by the number.

        Raises:
            TypeError: If ``other`` is not an actual number.
        """
        if not isinstance(other, Number):
            raise TypeError(f"Expected a number, got {type(other)!r}")

        return COOArray(
            data=other * self.data,
            row=self.row,
            col=self.col,
            shape=self.shape,
            has_canonical_format=self.has_canonical_format,
        )


def coo_array(
    arg,
    shape,
    *,
    validate: ValidationMode = "metadata",
    canonical: bool | None = None,
) -> COOArray:
    """Construct a :class:`COOArray` from coordinate arrays.

    Accepts either a ``(data, (row, col))`` pair or an existing
    ``COOArray``. All array inputs are converted to ``mlx.core.array`` if they
    are not already.

    Args:
        arg: A ``(data, (row, col))`` pair where

            - *data*: non-zero values, shape ``(nnz,)``, dtype
              ``float32 | float16 | bfloat16 | complex64``.
            - *row*: row coordinates, shape ``(nnz,)``, dtype
              ``int32 | int64``.
            - *col*: column coordinates, shape ``(nnz,)``, same integer
              dtype as *row*.

            Alternatively, an existing :class:`COOArray` (returned unchanged if
            ``shape`` matches).

        shape: Matrix dimensions as a length-2 sequence ``(n_rows, n_cols)``.
        validate: Validation level, one of:

            - ``"metadata"`` *(default)*: checks ranks, lengths, and dtypes.
            - ``"full"`` / ``True``: also verifies coordinate bounds. May
              synchronize to host.
            - ``False`` / ``"none"``: skips all checks.

        canonical: Set to ``True`` to assert the coordinates are sorted and
            duplicate-free. Default ``None`` (not asserted).

    Returns:
        A :class:`COOArray` with the given buffers and shape.

    Raises:
        TypeError: If ``arg`` cannot be unpacked as ``(data, (row, col))``, or
            if dtype constraints are violated.
        ValueError: If shape or length constraints are violated.

    Example::

        import mlx.core as mx
        import mlx_sparse as ms

        data = mx.array([1.0, 2.0, 3.0, 2.0], dtype=mx.float32)
        row = mx.array([0, 0, 1, 0], dtype=mx.int32)
        col = mx.array([0, 1, 0, 0], dtype=mx.int32)

        # Two entries at (0, 0). Summed when converting to CSR.
        coo = ms.coo_array((data, (row, col)), shape=(2, 2))
        csr = coo.tocsr(canonical=True)  # sums duplicates
    """
    mode = normalize_validation_mode(validate)
    shape = normalize_shape(shape)

    if isinstance(arg, COOArray):
        array = arg
        if array.shape != shape:
            raise ValueError(
                f"COOArray shape mismatch: got {array.shape}, expected {shape}."
            )
        return array

    try:
        data, coords = arg
        row, col = coords
    except Exception as exc:
        raise TypeError("coo_array expects (data, (row, col)) or a COOArray.") from exc

    data = ensure_mx_array(data)
    row = ensure_mx_array(row)
    col = ensure_mx_array(col)

    if mode != "none":
        validate_coo_metadata(data, row, col, shape)
    if mode == "full":
        validate_coo_values(row, col, shape)

    return COOArray(
        data=data,
        row=row,
        col=col,
        shape=shape,
        has_canonical_format=bool(canonical),
    )

File: src/random/coo.cpp
// Copyright (c) 2026 The mlx-sparse contributors - All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "random/random.h"

#include <cstdint>
#include <memory>
#include <stdexcept>
#include <string>
#include <tuple>
#include <vector>

#include "common/common.h"
#include "mlx/allocator.h"
#include "mlx/backend/cpu/encoder.h"
#include "mlx/ops.h"
#include "mlx/primitives.h"
#include "random/common.h"

#ifdef _METAL_
#include "mlx/backend/metal/device.h"
#endif

namespace mlx_sparse {

namespace {

class RandomCooUnpackSortedKeys : public mx::Primitive {
public:
  RandomCooUnpackSortedKeys(mx::Stream stream, int64_t n_cols)
      : Primitive(stream), n_cols_(n_cols) {}

  void eval_cpu(const std::vector<mx::array> &inputs,
                std::vector<mx::array> &outputs) override;

  void eval_gpu(const std::vector<mx::array> &inputs,
                std::vector<mx::array> &outputs) override;

  const char *name() const override { return "RandomCooUnpackSortedKeys"; }

  bool is_equivalent(const mx::Primitive &other) const override {
    const auto &rhs = static_cast<const RandomCooUnpackSortedKeys &>(other);
    return n_cols_ == rhs.n_cols_;
  }

private:
  int64_t n_cols_;
};

std::tuple<mx::array, mx::array>
random_coo_unpack_sorted_keys(const mx::array &keys, int64_t n_cols,
                              mx::Dtype index_dtype, mx::Stream stream) {
  auto primitive = std::make_shared<RandomCooUnpackSortedKeys>(stream, n_cols);
  auto outputs =
      mx::array::make_arrays({mx::Shape{static_cast<int>(keys.size())},
                              mx::Shape{static_cast<int>(keys.size())}},
                             {index_dtype, index_dtype}, primitive, {keys});
  return {outputs[0], outputs[1]};
}

} // namespace

void RandomCooUnpackSortedKeys::eval_cpu(const std::vector<mx::array> &inputs,
                                         std::vector<mx::array> &outputs) {
  const auto &keys = inputs[0];
  auto &row = outputs[0];
  auto &col = outputs[1];
  row.set_data(mx::allocator::malloc(row.nbytes()));
  col.set_data(mx::allocator::malloc(col.nbytes()));

  auto &encoder = mx::cpu::get_command_encoder(stream());
  encoder.set_input_array(keys);
  encoder.set_output_array(row);
  encoder.set_output_array(col);
  encoder.dispatch([keys = mx::array::unsafe_weak_copy(keys),
                    row = mx::array::unsafe_weak_copy(row),
                    col = mx::array::unsafe_weak_copy(col),
                    n_cols = n_cols_]() mutable {
    const auto *keys_ptr = keys.data<int64_t>();
    const auto nnz = keys.size();
    if (row.dtype() == mx::int32) {
      auto *row_ptr = row.data<int32_t>();
      auto *col_ptr = col.data<int32_t>();
      for (size_t i = 0; i < nnz; ++i) {
        const auto linear = static_cast<uint64_t>(keys_ptr[i]);
        row_ptr[i] =
            static_cast<int32_t>(linear / static_cast<uint64_t>(n_cols));
        col_ptr[i] =
            static_cast<int32_t>(linear % static_cast<uint64_t>(n_cols));
      }
    } else {
      auto *row_ptr = row.data<int64_t>();
      auto *col_ptr = col.data<int64_t>();
      for (size_t i = 0; i < nnz; ++i) {
        const auto linear = static_cast<uint64_t>(keys_ptr[i]);
        row_ptr[i] =
            static_cast<int64_t>(linear / static_cast<uint64_t>(n_cols));
        col_ptr[i] =
            static_cast<int64_t>(linear % static_cast<uint64_t>(n_cols));
      }
    }
  });
}

#ifdef _METAL_
void RandomCooUnpackSortedKeys::eval_gpu(const std::vector<mx::array> &inputs,
                                         std::vector<mx::array> &outputs) {
  const auto &keys = inputs[0];
  auto &row = outputs[0];
  auto &col = outputs[1];
  row.set_data(mx::allocator::malloc(row.nbytes()));
  col.set_data(mx::allocator::malloc(col.nbytes()));

  auto &s = stream();
  auto &device = mx::metal::device(s.device);
  auto *lib = device.get_library("mlx_sparse", current_binary_dir());
  auto kernel_name = std::string("random_coo_unpack_sorted_keys_") +
                     index_kernel_suffix(row.dtype());
  auto *kernel = device.get_kernel(kernel_name, lib);

  auto &encoder = mx::metal::get_command_encoder(s);
  encoder.set_compute_pipeline_state(kernel);
  encoder.set_input_array(keys, 0);
  encoder.set_output_array(row, 1);
  encoder.set_output_array(col, 2);
  encoder.set_bytes(n_cols_, 3);
  const int64_t nnz = static_cast<int64_t>(keys.size());
  encoder.set_bytes(nnz, 4);

  auto threads = std::max<size_t>(keys.size(), 1);
  auto group = std::min(threads, kernel->maxTotalThreadsPerThreadgroup());
  encoder.dispatch_threads(MTL::Size(threads, 1, 1), MTL::Size(group, 1, 1));
}
#else
void RandomCooUnpackSortedKeys::eval_gpu(const std::vector<mx::array> &,
                                         std::vector<mx::array> &) {
  throw std::runtime_error(
      "random COO key unpack has no GPU implementation in this build.");
}
#endif

std::tuple<mx::array, mx::array>
random_coo_indices(const mx::array &key, int64_t n_rows, int64_t n_cols,
                   int64_t nnz, int index_dtype_bits, mx::StreamOrDevice s) {
  check_random_key(key);
  const auto index_dtype = random_index_dtype_from_bits(index_dtype_bits);
  check_random_shape(n_rows, n_cols, nnz, index_dtype);

  auto stream = mx::to_stream(s);
  auto key_contig = mx::contiguous(key, false, stream);
  if (nnz == 0) {
    return {mx::zeros(mx::Shape{0}, index_dtype, stream),
            mx::zeros(mx::Shape{0}, index_dtype, stream)};
  }
  auto keys = random_structural_keys(key_contig, n_rows, n_cols, nnz,
                                     /*csc=*/false, stream);
  auto sorted_keys = mx::sort(keys, stream);
  return random_coo_unpack_sorted_keys(sorted_keys, n_cols, index_dtype,
                                       stream);
}

} // namespace mlx_sparse
